### PR TITLE
[966] Match cap with allocation if can order

### PR DESCRIPTION
**Pull request by asmega**
### Context

- https://trello.com/c/bSKwhZUM/966-updating-the-allocation-should-update-cap-if-a-school-can-order

### Changes proposed in this pull request

- If a school order state is `can_order` and allocation is updated, the cap is also increased to match the allocation

### Guidance to review

- Login as support user
- Find a school that `can_order`
- Increase/decrease the allocation
- Cap value should match the allocation value entered
- Change school so they `can_order_for specific_circumstances` or `cannot_order`
- Update allocation value
- Cap value should not change and remain the same

### CODE REVIEW

**Comment by asmega** [comment]
hmm actually hang back, this might not actually send the cap update request to CC?

File: .rubocop.yml [hunk 1]
```diff
@@ -49,6 +49,9 @@ RSpec/MultipleExpectations:
 RSpec/ContextWording:
   Enabled: false
 
+RSpec/MessageSpies:
+  Enabled: false
```
Comment (asmega) [inline .rubocop.yml:53]: I had to do this so I can do this:
```ruby
expect(SchoolOrderStateAndCapUpdateService).to receive(:new).with(school: school, order_state: school.order_state, std_device_cap: 100, coms_device_cap: 0).and_call_original
```
Unless someone has any better ideas?

File: spec/services/allocation_updater_spec.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      expect(allocation.allocation).to be(100)
+      expect(allocation.cap).to be(100)
+    end
+
+    it 'calls notifies computacenter with cap updates' do
+      expect(SchoolOrderStateAndCapUpdateService).to receive(:new).with(school: school, order_state: school.order_state, std_device_cap: 100, coms_device_cap: 0).and_call_original
+
+      service.call
+
+      expect(mock_request).to have_received(:post!).twice
```
Comment (aldavidson) [inline spec/services/allocation_updater_spec.rb:37]: I don't think this expectation belongs in this spec - as far as testing this service is concerned, it shouldn't need to know about the internals of what the `SchoolOrderStateAndCapUpdateService` does when you call `.update!`, just that an instance gets created with the right params and `.update!` gets called on it. That can be tested with a mock/double, which should mean you don't need to disable the RSpec MessageSpies rubocop option or mock the HTTP request.

File: spec/services/allocation_updater_spec.rb [hunk 3]
```diff
@@ -0,0 +1,84 @@
+require 'rails_helper'
+
+RSpec.describe AllocationUpdater do
+  let(:mock_request) { instance_double(Computacenter::OutgoingAPI::CapUpdateRequest, timestamp: Time.zone.now, payload_id: '123456789') }
```
Comment (aldavidson) [inline spec/services/allocation_updater_spec.rb:4]:
```suggestion
  let(:mock_update_service) { instance_double(SchoolOrderStateAndCapUpdateService) }

  before do
    allow(mock_update_service).to receive(:update!)
  end
```

File: spec/services/allocation_updater_spec.rb [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    before do
+      create(:school_device_allocation,
+             :with_std_allocation,
+             allocation: 10,
+             cap: 10,
+             school: school)
+
+      allow(Computacenter::OutgoingAPI::CapUpdateRequest).to receive(:new).and_return(mock_request)
+      allow(mock_request).to receive(:post!)
```
Comment (aldavidson) [inline spec/services/allocation_updater_spec.rb:22]:
```suggestion
```

File: spec/services/allocation_updater_spec.rb (hunk 2 above)
Comment (aldavidson) [inline spec/services/allocation_updater_spec.rb:37]:
```suggestion
      expect(SchoolOrderStateAndCapUpdateService).to receive(:new).with(school: school, order_state: school.order_state, std_device_cap: 100, coms_device_cap: 0).and_call_original

      service.call

      expect(mock_request).to have_received(:post!).twice      allow(SchoolOrderStateAndCapUpdateService).to receive(:new).and_return(mock_update_service)
      service.call

      expect(SchoolOrderStateAndCapUpdateService).to have_received(:new).with(school: school, order_state: school.order_state, std_device_cap: 100, coms_device_cap: 0)
      expect(mock_update_service).to have_received(:update!)
```

Comment (aldavidson) [inline spec/services/allocation_updater_spec.rb:37]:
```suggestion
            allow(SchoolOrderStateAndCapUpdateService).to receive(:new).and_return(mock_update_service)
      service.call

      expect(SchoolOrderStateAndCapUpdateService).to have_received(:new).with(school: school, order_state: school.order_state, std_device_cap: 100, coms_device_cap: 0)
      expect(mock_update_service).to have_received(:update!)
```

File: spec/services/allocation_updater_spec.rb [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  end
+
+  context 'allocation does not exist and school can order full allocation' do
+    let(:school) { create(:school, :in_lockdown) }
+
+    before do
+      allow(Computacenter::OutgoingAPI::CapUpdateRequest).to receive(:new).and_return(mock_request)
+      allow(mock_request).to receive(:post!)
+    end
+
```
Comment (aldavidson) [inline spec/services/allocation_updater_spec.rb:48]:
```suggestion
```

File: spec/services/allocation_updater_spec.rb [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+      expect(allocation.cap).to be(100)
+      expect(allocation).to be_persisted
+    end
+
+    it 'calls notifies computacenter with cap updates' do
+      expect(SchoolOrderStateAndCapUpdateService).to receive(:new).with(school: school, order_state: school.order_state, std_device_cap: 100, coms_device_cap: 0).and_call_original
+
+      service.call
+
+      expect(mock_request).to have_received(:post!).twice
```
Comment (aldavidson) [inline spec/services/allocation_updater_spec.rb:62]:
```suggestion
      allow(SchoolOrderStateAndCapUpdateService).to receive(:new).and_return(mock_update_service)
      service.call

      expect(SchoolOrderStateAndCapUpdateService).to have_received(:new).with(school: school, order_state: school.order_state, std_device_cap: 100, coms_device_cap: 0)
      expect(mock_update_service).to have_received(:update!)
```

File: .rubocop.yml (hunk 1 above)
Comment (aldavidson) [reply .rubocop.yml:53]: I'd prefer not to disable the Rubocops if possible, they do usually indicate that something could be better. 
In this case I think it's an expectation that doesn't need to be in the spec - see my [comment on the expectation](https://github.com/DFE-Digital/get-help-with-tech/pull/782/files#r519729286) and there's a batch of suggested changes that address it

**Review by aldavidson — CHANGES_REQUESTED** [review]
Hi Phil, looks good, I've made a batch of suggestions to the spec that mean you don't need to disable the MessageSpies.

**Comment by asmega** [comment]
@aldavidson The rubocop rule has been put back and your suggested changes should now be present

**Review by aldavidson — APPROVED** [review]
:+1: LGTM

**Pull request merged**